File: Python/Unique Binary Search Trees.py
class Solution:
    # @return an integer
    def numTrees(self, n):
        dp = []
        dp.append(1);
        for i in range(1, 15):
            res = 0;
            for j in range(1, i+1):
                 res += dp[j-1] * dp[i-j];
            dp.append(res);
        return dp[n];

File: C++/Unique-Binary-Search-Trees.cpp
class Solution {
public:
    int numTrees(int n) {
        if (n <= 1) return 1;
        vector<int> catalan(n + 1);
        catalan[0] = catalan[1] = 1;
        for (int i = 2; i <= n; i ++) {
            int cur = 0;
            for (int j = 0; j < i; j ++) {
                cur += catalan[j] * catalan[i - 1 - j];
            }
            catalan[i] = cur;
        }
        return catalan[n];
    }
};